### PR TITLE
Create listtablets shell command. Closes #1317

### LINKED ISSUE #1317: Provide single-line output of tablet info.
Copied from ACCUMULO-2873 - It would be very useful to have a utility that generates single line tablet info.  The output of this could be fed to sort, awk, grep, etc in order to answer questions like which tablets have the most files.The output could look something like the following
 
```
$accumulo admin listTablets --table bigTable3
 #files #walogs #entries #size #status #location #tableid #endrow
 6 2 40,001 50M ASSIGNED 10.1.9.9 4:9997[abc]  3 admin
 3 1 50,002 40M ASSIGNED 10.1.9.9 5:9997[abc]  3 helpful
```

All of the information can be obtained by scanning the metadata table and looking into zookeeper. Could possibly contact tablet servers to get info about entires in memory.

### PR DESCRIPTION
**Pull request by milleruntime**
* New command for debugging tablets called listtablets
* Added getLiveTServers() to TabletMetadata for generating a list of
tservers that currently have a lock in ZK, similar to master.
* The list of live tservers is passed to TabletMetadata in order to get
the current state of a tablet
* Command will print one line for every tablet in a table
* Created TabletMetadataIT for testing getLiveTServers()

Co-authored-by: EdColeman <dev1@etcoleman.com>

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 1]
```diff
@@ -435,4 +445,34 @@ static TabletMetadata create(String id, String prevEndRow, String endRow) {
...
+  public static synchronized Set<TServerInstance> getLiveTServers(ClientContext context) {
+    final Set<TServerInstance> currentServers = new HashSet<>();
+
+    final String path = context.getZooKeeperRoot() + Constants.ZTSERVERS;
+
+    HashSet<String> all = new HashSet<>(context.getZooCache().getChildren(path));
+
+    for (String zPath : all) {
+      checkServer(context, currentServers, path, zPath);
+    }
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:458]: This pattern of populating `currentServers` inside the other method, `checkServer`, after looping over `all` can be improved with something like:

```java
  var children = context.getZooCache().getChildren(path);
  var liveTservers = children.map(TabletMetadata::checkServer).filter(Optional::isPresent).collect(Collectors.toSet());
  log(liveTservers);
  return liveTservers;
```

This requires `checkServer` to be modified to return either `Optional.of(instance)` or `Optional.empty()`, but I think it's more readable.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  protected int defaultTimeoutSeconds() {
+    return 120;
+  }
+
+  private static class Callback implements MiniClusterConfigurationCallback {
+    @Override
+    public void configureMiniCluster(MiniAccumuloConfigImpl cfg, Configuration conf) {
+      cfg.setNumTservers(NUM_TSERVERS);
+    }
+  }
```
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java:70]: I feel like we have some sort of ConfigurableMacIT that is our normal way of doing this. Will that work here instead of this callback?

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 1 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:458]: I disagree.  To me what you have is longer and less readable.  I will rename all to children though.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java (hunk 2 above)
Comment (milleruntime) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java:70]: I don't know... despite having figured it out multiple times before, I still never know which IT to use... SharedMiniClusterBase, ConfigurableMacBase or AccumuloClusterHarness.  I just no not to use AccumuloITBase if you want a MiniCluster.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 1 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:458]: The main point I find difficult to read is the pass by reference of the data structure to a different method that is responsible for updating the data structure, conditionally. That kind of pattern can be hard to follow.

So, instead of:
```java
  methodToUpdateDataStructure(x1, x2, x3, dataStructure)
```

You can do:
```java
  var obj = shouldAddToDataStructure(x1, x2, x3);
  if (obj != null)
    dataStructure.add(obj);
```

So, if the part of my suggestion that was less readable to you was the stream stuff, here's a non-stream version:

```suggestion
    var currentServers = new HashSet<TServerInstance>();
    for (var child : context.getZooCache().getChildren(context.getZooKeeperRoot() + Constants.ZTSERVERS)) {
      var instance = checkServer(child);
      if (instance != null)
        currentServers.add(instance);
    }
```

I just thought the stream version was better, but honestly, anything is better than the pass-by-reference mutable data structure pattern, IMO.

If you prefer the current implementation, that's fine too. I don't feel very strongly about it. It was just something I noticed and thought could be improved.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:458]: I agree that mutating a data structure passed as a param can be deceptive.  For me personally, I typically find streams to be less readable.  I was just using what was there to start with since some of the code used in the locking I don't fully understand.  I do like your non-stream suggestion, I will use that.

File: test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java (hunk 2 above)
Comment (ctubbsii) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java:70]: I always have to look at existing classes to see what they are doing, and then I look at the type-hierarchy just to be sure. I think ConfigurableMacBase is appropriate here.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 1 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java]: @ctubbsii I think I came up with a nice combo of ideas in ba8fb27

File: test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java (hunk 2 above)
Comment (milleruntime) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java]: This made me think there are some ITs that could save on some code and use ConfigurableMacBase as well.  I will do that as a follow on.

**Comment by milleruntime** [comment]
@ctubbsii I think this PR good to go.  Did you have any other feedback before I merge?

**Comment by ctubbsii** [comment]
> @ctubbsii I think this PR good to go. Did you have any other feedback before I merge?

No. Go ahead.

**Comment by milleruntime** [comment]
FYI here is an example of output from the command using some test data:
<pre>
root@uno> listtablets -t test_ingest -h
2021-01-04T15:12:47,663 [Shell.audit] INFO : root@uno> listtablets -t test_ingest -h
NUM  TABLET_DIR      FILES WALS  ENTRIES   SIZE      STATUS     LOCATION                       ID    START (Exclusive)    END                 
TABLE: test_ingest
1    t-0000007       1     0            60       552 HOSTED     CURRENT:ip-10-113-12-25:9997   2     -INF                 row_0000000005      
2    t-0000006       1     0           500     2.71K HOSTED     CURRENT:ip-10-113-12-25:9997   2     row_0000000005       row_0000000055      
3    t-0000008       1     0         5.00K    24.74K HOSTED     CURRENT:ip-10-113-12-25:9997   2     row_0000000055       row_0000000555      
4    default_tablet  1     0         4.44K    22.01K HOSTED     CURRENT:ip-10-113-12-25:9997   2     row_0000000555       +INF                     
root@uno> listtablets -t accumulo.metadata
2021-01-04T15:13:21,750 [Shell.audit] INFO : root@uno> listtablets -t accumulo.metadata
NUM  TABLET_DIR      FILES WALS  ENTRIES   SIZE      STATUS     LOCATION                       ID    START (Exclusive)    END                 
TABLE: accumulo.metadata
1    table_info      2     0     7         524       HOSTED     CURRENT:ip-10-113-12-25:9997   !0    -INF                 ~                   
2    default_tablet  0     0     0         0         HOSTED     CURRENT:ip-10-113-12-25:9997   !0    ~                    +INF                
</pre>

File: test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java (hunk 2 above)
Comment (milleruntime) [reply test/src/main/java/org/apache/accumulo/test/functional/TabletMetadataIT.java]: So in another branch, I modified BulkNewIT to use ConfigurableMacBase (it is currently using SharedMiniClusterBase) and the test went from being around 30 seconds long to over 4 minutes.  I am not sure of the drastic difference between the two base classes but it seemed odd to me.
